### PR TITLE
feat!: Rule Tester checks for missing placeholder data in the message

### LINKED ISSUE #18016: Change Request: Always check data for message in rule tester
### ESLint version

current

### What problem do you want to solve?

Currently the rule tester only checks whether placeholders in messages are replaceable if the `data` property is set.
As a result there may be non-replaced placeholders even if this could be caught in the rule teser.
In other words the additional check can assure rule authors that all placeholders for their messages are available.
With the additional check the following can be caught:
```javascript
// Rule
export default {
	messages: {
		id: "{{ expected }}"
	},
	create(context) {
		context.report({ messageId: "id", data: { actual: true } });
	},
}

// Rule Tester
ruleTester.run("Rule", rule, {
	valid: [],
	invalid: [{ messageId: "id" }]
});
```
This is especially helpful as with the upcoming stricter rule tests as `message` cannot be set with `messageId` to check whether the placeholders are correctly replaced.

### What do you think is the correct solution?

Extend the logic for error / suggestion matcher to check whether all placeholders from the referenced message are available if there is no `data` property in the matcher.
This should also check that the `data` properties are not nullish.

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

I proposed this as a part of stricter rule tester testing in #17654 .

### PR DESCRIPTION
**Pull request by DMartens**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain: Rule Tester

#### What changes did you make? (Give an overview)
Fixes #18016.
Added a check for errors and suggestions which provide a `messageId` but not specify a `data` property.
This check makes sure that the replaced message has no placeholders as the required `data` properties were provided.

#### Is there anything you'd like reviewers to focus on?
- I copied the regular expression from `interpolate.js`; maybe the regular expression should be exported?
- Currently the check is only performed if there is no `data` property in the error. Maybe this check should run before as currently the `data` check does two things: detect missing placeholder data and checks whether the values match
- Maybe add a mention for suggestions that the `data` is not shared with the top level `data` ?

### CODE REVIEW

File: tests/lib/rule-tester/rule-tester.js [hunk 1]
```diff
@@ -2157,6 +2167,24 @@ describe("RuleTester", () => {
             });
         });
 
+        it("should pass with valid suggestions (tested using messageIds and data)", () => {
```
Comment (bmish) [inline tests/lib/rule-tester/rule-tester.js:2170]: Is this title correct?

File: tests/lib/rule-tester/rule-tester.js [hunk 2]
```diff
@@ -1906,6 +1907,15 @@ describe("RuleTester", () => {
         }, "Hydrated message \"Avoid using variables named 'notFoo'.\" does not match \"Avoid using variables named 'foo'.\"");
     });
 
+    it("should assert match between resulting message output if messageId and data provided in both test and result", () => {
```
Comment (bmish) [inline tests/lib/rule-tester/rule-tester.js:1910]: Is this title correct?

File: lib/rule-tester/rule-tester.js [hunk 3]
```diff
@@ -304,6 +304,17 @@ function throwForbiddenMethodError(methodName, prototype) {
...
 }
 
+const messagePlaceholderRegex = /\{\{([^{}]+?)\}\}/gu;
+
+/**
+ * Extracts names of missing {{ placeholders }} from the reported message.
+ * @param   {string} message Reported message
+ * @returns {string[]} Array of missing placeholder names
+ */
+function getMissingMessagePlaceholders(message) {
```
Comment (bmish) [inline lib/rule-tester/rule-tester.js:314]: Does the concept of "missing" actually come into play in this function? Or is just looking for placeholders in the message? If it's just looking for placeholders, which happen to be the missing ones in the context where it's called, then the function could be more generic by removing "missing" from its name.

File: lib/rule-tester/rule-tester.js [hunk 4]
```diff
@@ -304,6 +304,17 @@ function throwForbiddenMethodError(methodName, prototype) {
     };
 }
 
+const messagePlaceholderRegex = /\{\{([^{}]+?)\}\}/gu;
```
Comment (bmish) [inline lib/rule-tester/rule-tester.js:307]: I do think this regexp should be exported so it can be used in multiple places.

File: lib/rule-tester/rule-tester.js [hunk 5]
```diff
@@ -1012,6 +1023,13 @@ class RuleTester {
                                     rehydratedMessage,
                                     `Hydrated message "${rehydratedMessage}" does not match "${message.message}"`
                                 );
+                            } else {
+                                const missing = getMissingMessagePlaceholders(message.message);
+
+                                assert.ok(
+                                    missing.length === 0,
+                                    `The reported message has missing placeholders: ${missing.join(", ")}. Please provide them via the 'data' property in the context.report call`
```
Comment (bmish) [inline lib/rule-tester/rule-tester.js:1031]:
```suggestion
                                    `The reported message has missing placeholders: ${missing.join(", ")}. Please provide them via the 'data' property in the context.report() call.`
```

File: lib/rule-tester/rule-tester.js [hunk 6]
```diff
@@ -1105,6 +1123,13 @@ class RuleTester {
                                                     rehydratedDesc,
                                                     `${suggestionPrefix} Hydrated test desc "${rehydratedDesc}" does not match received desc "${actualSuggestion.desc}".`
                                                 );
+                                            } else {
+                                                const missing = getMissingMessagePlaceholders(actualSuggestion.desc);
+
+                                                assert.ok(
+                                                    missing.length === 0,
+                                                    `The message of the suggestion has missing placeholders: ${missing.join(", ")}. Please provide them via the 'data' property for the suggestion in the context.report call`
```
Comment (bmish) [inline lib/rule-tester/rule-tester.js:1131]:
```suggestion
                                                    `The message of the suggestion has missing placeholders: ${missing.join(", ")}. Please provide them via the 'data' property for the suggestion in the context.report() call.`
```

File: tests/lib/rule-tester/rule-tester.js [hunk 7]
```diff
@@ -1897,6 +1897,7 @@ describe("RuleTester", () => {
             invalid: [{ code: "foo", errors: [{ messageId: "avoidFoo" }] }]
         });
     });
+
```
Comment (bmish) [inline tests/lib/rule-tester/rule-tester.js:1900]: Do we need to worry about the pathological edge case where the placeholder data contains another placeholder? It would hit your new exception unless we could escape it somehow. At first, I figured this wouldn't be a realistic situation, but maybe it could be if someone was writing a lint rule that was linting placeholders themselves. However we decide to handle it, it would be good to document the expected behavior in the form of a test case.

File: tests/lib/rule-tester/rule-tester.js [hunk 8]
```diff
@@ -1906,6 +1907,24 @@ describe("RuleTester", () => {
         }, "Hydrated message \"Avoid using variables named 'notFoo'.\" does not match \"Avoid using variables named 'foo'.\"");
     });
 
+    it("should throw if the message has missing placeholders when data is not specified", () => {
+        assert.throws(() => {
+            ruleTester.run("foo", require("../../fixtures/testers/rule-tester/messageId").withMissingData, {
+                valid: [],
+                invalid: [{ code: "foo", errors: [{ messageId: "avoidFoo" }] }]
+            });
+        }, "The reported message has missing placeholders: name. Please provide them via the 'data' property in the context.report call");
```
Comment (JoshuaKGoldberg) [inline tests/lib/rule-tester/rule-tester.js:1916]: [Feature] Maybe I'm being too nitpicky here, but I think it'd be good to adjust the grammar to handle single vs. plural cases.

```suggestion
        }, "The reported message has a missing placeholder: name. Please provide it via the 'data' property in the context.report call");
```

File: tests/lib/rule-tester/rule-tester.js
Comment (JoshuaKGoldberg) [inline tests/lib/rule-tester/rule-tester.js:1]: [Testing] Could you also add tests for multiple missing placeholders? The `.join(", ")`s should be exercised to make sure they stay working in the future, I think.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Seems reasonable and straightforward to me! 🙌 

+1 to everything bmish said. Left a couple small comments too.

File: lib/rule-tester/rule-tester.js [hunk 9]
```diff
@@ -1012,6 +1023,13 @@ class RuleTester {
                                     rehydratedMessage,
                                     `Hydrated message "${rehydratedMessage}" does not match "${message.message}"`
                                 );
+                            } else {
```
Comment (mdjermanovic) [inline lib/rule-tester/rule-tester.js:1026]: I think we should check this even if there is `data` property in the test error.

It would catch cases like this:

```js
messages: {
    unexpected: "Unexpected {{nam}}." // typo name -> nam
}

context.report({
    node,
    messageId: "unexpected",
    data: { name }
});

// this test is passing, although the message is  "Unexpected {{nam}}." literally.
{
    code: "alert(foo)",
    errors: [{ messageId: "unexpected", data: { name: "alert" }]
}
```

File: lib/rule-tester/rule-tester.js [hunk 10]
```diff
@@ -304,6 +304,17 @@ function throwForbiddenMethodError(methodName, prototype) {
     };
 }
 
+const messagePlaceholderRegex = /\{\{([^{}]+?)\}\}/gu;
+
+/**
+ * Extracts names of missing {{ placeholders }} from the reported message.
+ * @param   {string} message Reported message
+ * @returns {string[]} Array of missing placeholder names
```
Comment (bmish) [inline lib/rule-tester/rule-tester.js:312]:
```suggestion
 * Extracts names of {{ placeholders }} from the reported message.
 * @param   {string} message Reported message
 * @returns {string[]} Array of placeholder names
```

**Comment by DMartens** [comment]
Implemented all the suggestions except for the potential exporting of the regular expression.
This could be done in different ways:
- export directly from `interpolate.js`
- create a new utility file in `lib/shared/` which both `interpolate.js` and the rule tester uses
Further instead of exporting the regular expression, we could export a function returning an iterator for the regular expression matches.

File: tests/lib/rule-tester/rule-tester.js [hunk 11]
```diff
@@ -1906,6 +1907,42 @@ describe("RuleTester", () => {
...
+        }, "The reported message has the missing placeholders: 'type', 'name'. Please provide them via the 'data' property in the context.report() call.");
+    });
+
+    it("should throw if the data in the message contains placeholders", () => {
+        assert.throws(() => {
+            ruleTester.run("foo", require("../../fixtures/testers/rule-tester/messageId").withPlaceholdersInData, {
+                valid: [],
+                invalid: [{ code: "foo", errors: [{ messageId: "avoidFoo" }] }]
+            });
+        }, "The reported message has a missing placeholder 'placeholder'. Please provide them via the 'data' property in the context.report() call.");
```
Comment (bmish) [inline tests/lib/rule-tester/rule-tester.js:1943]: This error message seems misleading since, as @mdjermanovic said, we don't want to suggest that recursive/multipass is allowed. If there was someway to escape placeholders in `data`, that could solve it.

**Comment by mdjermanovic** [comment]
> Also now that the missing placeholders are also checked when `data` is specified, there is a case in [no-restricted-syntax](https://github.com/eslint/eslint/blob/ce838adc3b673e52a151f36da0eedf5876977514/tests/lib/rules/no-restricted-syntax.js#L123) which uses a placeholder in the provided placeholder data.
> How should this be handled (e.g. should this be supported (e.g. via a recursive interpolate)?
> Otherwise we could check the passed `data` properties from the error and do not include the detected missing placeholders.

If there is test `data`, we could just check if all placeholders from the meta message have corresponding properties in the test `data` object. With this check, and the already existing check whether the calculated test message text matches the actual message text, I think there's even no need to look for missing placeholders in the actual message text.

If there isn't test `data`, I think we can't be 100% sure that the placeholder wasn't replaced (theoretically, it could have been replaced with the exact same text).

File: tests/lib/rule-tester/rule-tester.js (hunk 11 above)
Comment (mdjermanovic) [reply tests/lib/rule-tester/rule-tester.js:1943]: When only `messageId` is specified in the test case, I think we could just check if the meta message really has those placeholders we found as missing. While theoretically the same name can appear as a placeholder and in data, that seems like an extreme edge case.

**Review by nzakas — CHANGES_REQUESTED** [review]
Also, please update the migrate-to-9.0.0.md file to mention this change.

File: lib/rule-tester/rule-tester.js [hunk 12]
```diff
@@ -304,6 +305,40 @@ function throwForbiddenMethodError(methodName, prototype) {
...
+
+    if (data === void 0) {
+        const available = getMessagePlaceholders(raw);
+
+        return existing.filter(name => available.includes(name));
+    }
+
+    const introduced = new Set(Object.values(data).flatMap(value => (typeof value === "string" ? getMessagePlaceholders(value) : [])));
+
+    return existing.filter(name => !introduced.has(name));
```
Comment (mdjermanovic) [inline lib/rule-tester/rule-tester.js:339]:
```suggestion
    let existing = getMessagePlaceholders(message);

    //  no potential placeholders were found in the reported message, so there's nothing further to check
    if (existing.length === 0) {
        return existing;
    }

    const original = getMessagePlaceholders(raw);

    // filter out those that don't appear in the original template message as they're certainly not placeholders
    existing = existing.filter(name => original.includes(name));

    if (data) {
        const provided = Object.keys(data);

        // since the hydrated message texts are compared, we can ignore those for which the data is provided in the test
        existing = existing.filter(name => !provided.includes(name));
    }

    return existing;
```

I think this check would be more robust if we always exclude those that don't appear in the meta message and those for which the data is provided in the test.

**Comment by mdjermanovic** [comment]
Also, there are merge conflicts in `docs/src/use/migrate-to-9.0.0.md`.

File: lib/rule-tester/rule-tester.js (hunk 12 above)
Comment (DMartens) [reply lib/rule-tester/rule-tester.js]: I refactored this method in another way and also kept the case for adding placeholders via data properties (see `no-restricted-syntax` case).

File: lib/rule-tester/rule-tester.js [hunk 13]
```diff
@@ -997,6 +1032,14 @@ class RuleTester {
                                 error.messageId,
                                 `messageId '${message.messageId}' does not match expected messageId '${error.messageId}'.`
                             );
+
+                            const missingPlaceholders = getMissingMessagePlaceholders(message.message, rule.meta.messages[message.messageId], error.data);
+
+                            assert.ok(
+                                missingPlaceholders.length === 0,
+                                `The reported message has ${missingPlaceholders.length > 1 ? `the missing placeholders: ${missingPlaceholders.map(name => `'${name}'`).join(", ")}` : `a missing placeholder '${missingPlaceholders[0]}'`}. Please provide them via the 'data' property in the context.report() call.`
```
[the remark this answers is not in the document]
Comment (DMartens) [reply lib/rule-tester/rule-tester.js]: I used the term `missing` as this is the cause of the issue. That the placeholder cannot be substituted is the result of that. But I also don't really like missing and I cannot come up with a better name for it, so I changed it in the messages and variable names.

File: docs/src/use/migrate-to-9.0.0.md [hunk 14]
```diff
@@ -546,6 +546,7 @@ In order to aid in the development of high-quality custom rules that are free fr
 1. **Suggestion messages must be unique.** Because suggestions are typically displayed in an editor as a dropdown list, it's important that no two suggestions for the same lint problem have the same message. Otherwise, it's impossible to know what any given suggestion will do. This additional check runs automatically.
 1. **Suggestions must generate valid syntax.** In order for rule suggestions to be helpful, they need to be valid syntax. `RuleTester` now parses the output of suggestions using the same language options as the `code` value and throws an error if parsing fails.
 1. **Test cases must be unique.** Identical test cases can cause confusion and be hard to detect manually in a long test file. Duplicates are now automatically detected and can be safely removed.
+1. **Messages cannot have missing placeholders.** The rule tester now checks if there are `{{ placeholder }}` still in the message as they were not passed via `data` in the respective `context.report`.
```
Comment (mdjermanovic) [inline docs/src/use/migrate-to-9.0.0.md:549]:
```suggestion
1. **Messages cannot have unsubstituted placeholders.** The rule tester now checks if there are `{{ placeholder }}` still in the message as they were not passed via `data` in the respective `context.report`.
```

File: lib/rule-tester/rule-tester.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/rule-tester/rule-tester.js]: I think this is a grammar issue: the placeholder isn't missing, it's the value to use for the placeholder that's missing. Another option would be "missing placeholder value".

**Comment by mdjermanovic** [comment]
@DMartens can you fix merge conflicts? We'd like to include this in today's v9.0.0-beta.0 release.

File: lib/rule-tester/rule-tester.js [hunk 15]
```diff
@@ -304,6 +305,41 @@ function throwForbiddenMethodError(methodName, prototype) {
...
+    if (unsubstituted.length === 0) {
+        return [];
+    }
+
+    // Remove false positives by only counting placeholders in the raw message, which were not provided in the data matcher or added with a data property
+    const known = getMessagePlaceholders(raw);
+    const provided = Object.keys(data);
+    const added = new Set(Object.values(data).flatMap(value => (typeof value === "string" ? getMessagePlaceholders(value) : [])));
+
+    return unsubstituted.filter(name => known.includes(name) && !provided.includes(name) && !added.has(name));
```
Comment (mdjermanovic) [inline lib/rule-tester/rule-tester.js:340]: I'm not sure if we need `&& !added.has(name)` here. Is there a case where it removes more false positives? I can only think of cases where it might introduce false negatives (though, those are extreme edge cases).

File: docs/src/use/migrate-to-9.0.0.md [hunk 16]
```diff
@@ -546,6 +546,7 @@ In order to aid in the development of high-quality custom rules that are free fr
 1. **Suggestion messages must be unique.** Because suggestions are typically displayed in an editor as a dropdown list, it's important that no two suggestions for the same lint problem have the same message. Otherwise, it's impossible to know what any given suggestion will do. This additional check runs automatically.
 1. **Suggestions must generate valid syntax.** In order for rule suggestions to be helpful, they need to be valid syntax. `RuleTester` now parses the output of suggestions using the same language options as the `code` value and throws an error if parsing fails.
 1. **Test cases must be unique.** Identical test cases can cause confusion and be hard to detect manually in a long test file. Duplicates are now automatically detected and can be safely removed.
+1. **Messages cannot have unsubstituted placeholders.** The rule tester now checks if there are `{{ placeholder }}` still in the message as they were not passed via `data` in the respective `context.report`.
```
Comment (mdjermanovic) [inline docs/src/use/migrate-to-9.0.0.md:549]:
```suggestion
1. **Messages cannot have unsubstituted placeholders.** The rule tester now checks if there are `{{ placeholder }}` still in the message as their values were not passed via `data` in the respective `context.report`.
```

File: lib/rule-tester/rule-tester.js (hunk 15 above)
Comment (DMartens) [reply lib/rule-tester/rule-tester.js:340]: For cases like in [no-restricted-syntax](https://github.com/eslint/eslint/blob/15c143f96ef164943fd3d39b5ad79d9a4a40de8f/tests/lib/rules/no-restricted-syntax.js#L123) where a case "adding" a new placeholder  by using placeholders via a data value (not the same name as the data property)

Comment (mdjermanovic) [reply lib/rule-tester/rule-tester.js:340]: > For cases like in [no-restricted-syntax](https://github.com/eslint/eslint/blob/15c143f96ef164943fd3d39b5ad79d9a4a40de8f/tests/lib/rules/no-restricted-syntax.js#L123) where a case "adding" a new placeholder by using placeholders via a data value (not the same name as the data property)

Wouldn't that be already filtered out by `known.includes(name) && !provided.includes(name)`?

File: docs/src/use/migrate-to-9.0.0.md [hunk 17]
```diff
@@ -495,7 +495,7 @@ As announced in our [blog post](/blog/2023/10/flat-config-rollout-plans/), the t
 
 **To address:** Update your rule tests to use the new `RuleTester`. To do so, here are some of the common changes you'll need to make:
 
-* **Be aware of new defaults for `ecmaVersion` and `sourceType`.** By default, `RuleTester` uses the flat config default of `ecmaVersion: "latest"` and `sourceType: "module"`. This may cause some tests to break if they were expecting the old default of `ecmaVersion: 5` and `sourceType: "script"`. If you'd like to use the old default, you'll need to manually specify that in your `RuleTester` like this:
+* **Be aware of new defaults for `ecmaVersion` and `sourceType`.** By default, `RuleTester` uses the flat config default of `ecmaVersion: "latest"` and `sourceType: "module"`. This may cause some tests to break if their values were expecting the old default of `ecmaVersion: 5` and `sourceType: "script"`. If you'd like to use the old default, you'll need to manually specify that in your `RuleTester` like this:
```
Comment (mdjermanovic) [inline docs/src/use/migrate-to-9.0.0.md:498]:
```suggestion
* **Be aware of new defaults for `ecmaVersion` and `sourceType`.** By default, `RuleTester` uses the flat config default of `ecmaVersion: "latest"` and `sourceType: "module"`. This may cause some tests to break if they were expecting the old default of `ecmaVersion: 5` and `sourceType: "script"`. If you'd like to use the old default, you'll need to manually specify that in your `RuleTester` like this:
```

File: lib/rule-tester/rule-tester.js [hunk 18]
```diff
@@ -997,6 +1033,18 @@ class RuleTester {
...
+
+                            const unsubstitutedPlaceholders = getUnsubstitutedMessagePlaceholders(
+                                message.message,
+                                rule.meta.messages[message.messageId],
+                                error.data
+                            );
+
+                            assert.ok(
+                                unsubstitutedPlaceholders.length === 0,
+                                `The reported message has ${unsubstitutedPlaceholders.length > 1 ? `the unsubstituted placeholders: ${unsubstitutedPlaceholders.map(name => `'${name}'`).join(", ")}` : `a unsubstituted placeholder '${unsubstitutedPlaceholders[0]}'`}. Please provide them via the 'data' property in the context.report() call.`
```
Comment (mdjermanovic) [inline lib/rule-tester/rule-tester.js:1045]:
```suggestion
                                `The reported message has ${unsubstitutedPlaceholders.length > 1 ? `unsubstituted placeholders: ${unsubstitutedPlaceholders.map(name => `'${name}'`).join(", ")}` : `an unsubstituted placeholder '${unsubstitutedPlaceholders[0]}'`}. Please provide them via the 'data' property in the context.report() call.`
```

File: lib/rule-tester/rule-tester.js [hunk 19]
```diff
@@ -1096,6 +1144,18 @@ class RuleTester {
...
+
+                                            const missingPlaceholders = getUnsubstitutedMessagePlaceholders(
+                                                actualSuggestion.desc,
+                                                rule.meta.messages[expectedSuggestion.messageId],
+                                                expectedSuggestion.data
+                                            );
+
+                                            assert.ok(
+                                                missingPlaceholders.length === 0,
+                                                `The message of the suggestion has ${missingPlaceholders.length > 1 ? `the unsubstituted placeholders: ${missingPlaceholders.map(name => `'${name}'`).join(", ")}` : `a unsubstituted placeholder '${missingPlaceholders[0]}'`}. Please provide them via the 'data' property for the suggestion in the context.report() call.`
```
Comment (mdjermanovic) [inline lib/rule-tester/rule-tester.js:1156]:
```suggestion
                                                `The message of the suggestion has ${missingPlaceholders.length > 1 ? `unsubstituted placeholders: ${missingPlaceholders.map(name => `'${name}'`).join(", ")}` : `an unsubstituted placeholder '${missingPlaceholders[0]}'`}. Please provide them via the 'data' property for the suggestion in the context.report() call.`
```

File: tests/lib/rule-tester/rule-tester.js [hunk 20]
```diff
@@ -1906,6 +1907,65 @@ describe("RuleTester", () => {
...
+    });
+
+    it("should not throw an error for specifying non-string data values", () => {
+        assert.throws(() => {
+            ruleTester.run("foo", require("../../fixtures/testers/rule-tester/messageId").withSamePlaceholdersInData, {
+                valid: [],
+                invalid: [{ code: "foo", errors: [{ messageId: "avoidFoo", data: { name: 0 } }] }]
+            });
+        }, "Hydrated message \"Avoid using variables named '0'.\" does not match \"Avoid using variables named '{{ name }}'.\"");
+    });
```
Comment (mdjermanovic) [inline tests/lib/rule-tester/rule-tester.js:1967]: This is confusing because the test title says "should not throw an error" but then asserts that it does throw an error. Perhaps we need another fixture rule that passes `data: { name: 0 }`?

File: docs/src/use/migrate-to-9.0.0.md [hunk 21]
```diff
@@ -553,6 +553,7 @@ In order to aid in the development of high-quality custom rules that are free fr
 1. **Suggestions must generate valid syntax.** In order for rule suggestions to be helpful, they need to be valid syntax. `RuleTester` now parses the output of suggestions using the same language options as the `code` value and throws an error if parsing fails.
 1. **Test cases must be unique.** Identical test cases can cause confusion and be hard to detect manually in a long test file. Duplicates are now automatically detected and can be safely removed.
 1. **`filename` and `only` must be of the expected type.** `RuleTester` now checks the type of `filename` and `only` properties of test objects. If specified, `filename` must be a string value. If specified, `only` must be a boolean value.
+1. **Messages cannot have unsubstituted placeholders.** The rule tester now checks if there are `{{ placeholder }}` still in the message as their values were not passed via `data` in the respective `context.report`.
```
Comment (mdjermanovic) [inline docs/src/use/migrate-to-9.0.0.md:556]:
```suggestion
1. **Messages cannot have unsubstituted placeholders.** The rule tester now checks if there are {% raw %}`{{ placeholder }}`{% endraw %} still in the message as their values were not passed via `data` in the respective `context.report()` call.
```

We need to wrap this into `{% raw %}...{% endraw %}`, otherwise this syntax would have a meaning for tools we use to build the site. Here's how it currently looks in this PR's preview (note that the`{{ placeholder }}` text doesn't appear):

image

Also, context.report -> context.report() call.

Also, can you add `[#18016](https://github.com/eslint/eslint/issues/18016)` to the list of related issues on line 560 (GitHub doesn't allow me to leave suggestions there)?

File: lib/linter/interpolate.js [hunk 22]
```diff
@@ -9,20 +9,27 @@
...
+    },
+    interpolate(text, data) {
+        if (!data) {
+            return text;
+        }
 
-    // Substitution content for any {{ }} markers.
-    return text.replace(/\{\{([^{}]+?)\}\}/gu, (fullMatch, termWithWhitespace) => {
-        const term = termWithWhitespace.trim();
+        const matcher = module.exports.getPlaceholderMatcher();
```
Comment (nzakas) [inline lib/linter/interpolate.js:21]: To avoid confusion.
```suggestion
        const matcher = this.getPlaceholderMatcher();
```

**Review by nzakas — APPROVED** [review]
LGTM. Just one suggestion for clarity but otherwise good to go.

File: lib/linter/interpolate.js (hunk 22 above)
Comment (mdjermanovic) [reply lib/linter/interpolate.js:21]: `this` wouldn't work unless the consuming modules always import the whole object and call this function as a method through that object.

I think it would be best to simply declare and export two functions:

```js
function getPlaceholderMatcher() {
   // ...
}

function interpolate() {
    // .... just call getPlaceholderMatcher() as a function here.
}

module.exports = {
    getPlaceholderMatcher,
    interpolate
}
```

File: lib/rule-tester/rule-tester.js [hunk 23]
```diff
@@ -997,6 +1031,18 @@ class RuleTester {
...
+
+                            const unsubstitutedPlaceholders = getUnsubstitutedMessagePlaceholders(
+                                message.message,
+                                rule.meta.messages[message.messageId],
+                                error.data
+                            );
+
+                            assert.ok(
+                                unsubstitutedPlaceholders.length === 0,
+                                `The reported message has ${unsubstitutedPlaceholders.length > 1 ? `unsubstituted placeholders: ${unsubstitutedPlaceholders.map(name => `'${name}'`).join(", ")}` : `an unsubstituted placeholder '${unsubstitutedPlaceholders[0]}'`}. Please provide them via the 'data' property in the context.report() call.`
```
Comment (mdjermanovic) [inline lib/rule-tester/rule-tester.js:1043]:
```suggestion
                                `The reported message has ${unsubstitutedPlaceholders.length > 1 ? `unsubstituted placeholders: ${unsubstitutedPlaceholders.map(name => `'${name}'`).join(", ")}` : `an unsubstituted placeholder '${unsubstitutedPlaceholders[0]}'`}. Please provide the missing ${unsubstitutedPlaceholders.length > 1 ? "values" : "value"} via the 'data' property in the context.report() call.`
```

File: lib/rule-tester/rule-tester.js [hunk 24]
```diff
@@ -1096,6 +1142,18 @@ class RuleTester {
...
+
+                                            const missingPlaceholders = getUnsubstitutedMessagePlaceholders(
+                                                actualSuggestion.desc,
+                                                rule.meta.messages[expectedSuggestion.messageId],
+                                                expectedSuggestion.data
+                                            );
+
+                                            assert.ok(
+                                                missingPlaceholders.length === 0,
+                                                `The message of the suggestion has ${missingPlaceholders.length > 1 ? `unsubstituted placeholders: ${missingPlaceholders.map(name => `'${name}'`).join(", ")}` : `an unsubstituted placeholder '${missingPlaceholders[0]}'`}. Please provide them via the 'data' property for the suggestion in the context.report() call.`
```
Comment (mdjermanovic) [inline lib/rule-tester/rule-tester.js:1154]:
```suggestion
                                                `The message of the suggestion has ${missingPlaceholders.length > 1 ? `unsubstituted placeholders: ${missingPlaceholders.map(name => `'${name}'`).join(", ")}` : `an unsubstituted placeholder '${missingPlaceholders[0]}'`}. Please provide the missing ${missingPlaceholders.length > 1 ? "values" : "value"} via the 'data' property for the suggestion in the context.report() call.`
```

File: docs/src/use/migrate-to-9.0.0.md [hunk 25]
```diff
@@ -553,10 +553,11 @@ In order to aid in the development of high-quality custom rules that are free fr
 1. **Suggestions must generate valid syntax.** In order for rule suggestions to be helpful, they need to be valid syntax. `RuleTester` now parses the output of suggestions using the same language options as the `code` value and throws an error if parsing fails.
 1. **Test cases must be unique.** Identical test cases can cause confusion and be hard to detect manually in a long test file. Duplicates are now automatically detected and can be safely removed.
 1. **`filename` and `only` must be of the expected type.** `RuleTester` now checks the type of `filename` and `only` properties of test objects. If specified, `filename` must be a string value. If specified, `only` must be a boolean value.
+1. **Messages cannot have unsubstituted placeholders.** The rule tester now checks if there are {% raw %}`{{ placeholder }}` {% endraw %} still in the message as their values were not passed via `data` in the respective `context.report()` call.
```
Comment (snitin315) [inline docs/src/use/migrate-to-9.0.0.md:556]:
```suggestion
1. **Messages cannot have unsubstituted placeholders.** The `RuleTester` now checks if there are {% raw %}`{{ placeholder }}` {% endraw %} still in the message as their values were not passed via `data` in the respective `context.report()` call.
```

**Review by snitin315 — APPROVED** [review]
LGTM, thank you for your contribution.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Comment by mdjermanovic** [comment]
I've verified that all review comments have been addressed, so merging for the release.
